examples/vision/image_captioning.py

```diff
@@ -0,0 +1,486 @@
+        )
+        out_1 = self.layernorm_1(inputs + attention_output_1)
+
+        attention_output_2 = self.attention_2(
+            query=out_1, value=encoder_outputs, key=encoder_outputs, attention_mask=None
+        )
+        out_2 = self.layernorm_2(out_1 + attention_output_2)
+
+        proj_output = self.dense_proj(out_2)
+        proj_out = self.layernorm_3(out_2 + proj_output)
+        proj_out = self.dropout(proj_out, training=training)
+
+        preds = self.out(proj_out)
+        return preds
+
+    def get_causal_attention_mask(self, inputs):
+        input_shape = tf.shape(inputs)
+        batch_size, sequence_length = input_shape[0], input_shape[1]
+        i = tf.range(sequence_length)[:, tf.newaxis]
+        j = tf.range(sequence_length)
+        mask = tf.cast(i >= j, dtype="int32")
+        mask = tf.reshape(mask, (1, input_shape[1], input_shape[1]))
+        mult = tf.concat(
+            [tf.expand_dims(batch_size, -1), tf.constant([1, 1], dtype=tf.int32)],
+            axis=0,
+        )
+        return tf.tile(mask, mult)
+
+
+class ImageCaptioningModel(keras.Model):
+    def __init__(
+        self,
+        cnn_model,
+        encoder,
+        decoder,
+    ):
+        super(ImageCaptioningModel, self).__init__()
+        self.cnn_model = cnn_model
+        self.encoder = encoder
+        self.decoder = decoder
+        self.train_loss_tracker = keras.metrics.Mean(name="loss")
+        self.val_loss_tracker = keras.metrics.Mean(name="val_loss")
+
+    def _compute_loss(self, batch_data, training=True):
+        batch_img, batch_seq = batch_data
+        loss = 0
+
+        # 1. Get image embeddings
+        img_embed = self.cnn_model(batch_img)
+
+        # 2. Pass image embeddings to encoder
+        encoder_out = self.encoder(img_embed, training=training)
+
+        # 3. Pass each of the five captions one by one the decoder
+        # along with the encoder outputs and compute the loss for each caption
+        for i in range(5):
+            batch_seq_inp = batch_seq[:, i, :-1]
+            batch_seq_true = batch_seq[:, i, 1:]
+            batch_seq_pred = self.decoder(batch_seq_inp, encoder_out, training=training)
+            loss += self.loss(batch_seq_true, batch_seq_pred)
+        return loss
+
+    def train_step(self, batch_data):
+        # 1. Compute the loss
+        with tf.GradientTape() as tape:
+            loss = self._compute_loss(batch_data)
+
+        # 2. Get the list of all the trainable weights
+        train_vars = self.encoder.trainable_variables + self.decoder.trainable_variables
+
+        # 7. Get the gradients
+        grads = tape.gradient(loss, train_vars)
+
+        # 8. Update the trainable weights
+        self.optimizer.apply_gradients(zip(grads, train_vars))
+
+        # 9. Update the loss tracker and return the loss value
+        self.train_loss_tracker.update_state(loss)
+        return {"loss": self.train_loss_tracker.result()}
```

Comment:
Why not keep track of another metric like accuracy?

Reply:
Will update this